
<!-- README.md is generated from README.Rmd. Please edit that file -->

<img src="man/figures/ProtE.png" width="60%" />

One function to analyze them all! The Proteomics Eye (ProtE) establishes
an intuitive framework for the univariate analysis of label-free
proteomics data. By compiling all necessary data wrangling and
processing steps into the same function, ProtE automates all pairwise
statistical comparisons for a given categorical variable, returning to
the user performance quality metrics, measures to control for Type-I or
Type-II errors, and publication-ready visualizations.

ProtE is currently compatible with data generated by MaxQuant, DIA-NN
and Proteome Discoverer.

## How to install:

If you already have R Studio and Rtools in your PC skip to Step 4.

Step 1: Install R

To get started with R, first download and install the latest version of
R from the official CRAN website:

- Go to the [R download page](https://cran.r-project.org/).
- Select the appropriate version for your operating system:
  - **Windows**: Click on “Download R for Windows”.
  - **MacOS**: Click on “Download R for macOS”.
  - **Linux**: Follow the instructions based on your distribution.

Once downloaded, run the installer and follow the instructions to
complete the installation.

Step 2: Install RStudio

Next, you’ll need an Integrated Development Environment (IDE) to work
with R. The most popular IDE is RStudio.

- Go to the [RStudio download
  page](https://posit.co/download/rstudio-desktop/).
- Select “RStudio Desktop” and download the installer for your operating
  system.

Run the installer and follow the on-screen instructions to install
RStudio.

Step 3: Install RTools (For Windows Users)

RTools is necessary if you need to compile packages from source on
Windows, which is common when installing certain R packages.

- Go to the [RTools download
  page](https://cran.r-project.org/bin/windows/Rtools/).
- Download the version of RTools that matches your R version.
- Run the installer and follow the instructions.

Make sure to select the option that allows RTools to be added to your
system path during installation.

Step 4: Download the Package

Install the development version of ProtE from GitHub, along with its
Bioconductor dependencies:

    if (!require("devtools", quietly = TRUE)) {
      install.packages("devtools")}
      
    if (!require("BiocManager", quietly = TRUE)) {
      install.packages("BiocManager")}
      
    bioc_packages <- c("limma", "ComplexHeatmap", "fgsea")
    for (pkg in bioc_packages) {
      if (!require(pkg, character.only = TRUE, quietly = TRUE)) {
        BiocManager::install(pkg) }}

    devtools::install_github("yourusername/ProtE", dependencies = TRUE)

Then load its library with:


    library(ProtE)

One function to analyze them all! The Proteomics Eye (ProtE) establishes
an intuitive framework for the univariate analysis of label-free
proteomics data. By compiling all necessary data wrangling and
processing steps into the same function, ProtE automates all statistical
comparisons for a given categorical variable, returning to the user
performance quality metrics, measures to control for Type-I or Type-II
errors, and publication-ready visualizations.

ProtE is currently compatible with data generated by MaxQuant, DIA-NN
and Proteome Discoverer.

## Function inputs

ProtE features just one function ; `ProtE_analyse()`, which accepts as
input the specific use cases:

1.  The MaxQuant generated file ProteinGroups.txt or the same file
    transposed to .xlsx format
2.  Either of the two DIA-NN (or the FragPipe - DIANN) output files
    pg_matrix.tsv / unique_genes_matrix.tsv or the same files transposed
    to .xlsx format
3.  The Proteome Discoverer output file that contains all sample protein
    intensities/abundances in one Excel table
4.  OR separate Proteome Discoverer protein intensity files for each
    sample in .xlsx format

The input file for cases 1-3 should be parsed in the parameter `file`
While for case 4 the path to the folder(s) that contain the .xlsx files
for each sample, should be parsed in the parameter `pd_single_dir`.

A metadata table (.xlsx or .txt format) is also optional to be provided,
via the parameter `metadata_file`. It must containthe Sample names that
should match the naming inside the input table (or the naming of the
Excel files for case 4.) in the 1st column, and the experimental Groups
of comparison in the 2nd column. The next columns can include
information about covariates that will be taken into account in the
statistical analysis.

## Setting up the file path:

Assuming a MaxQuant quantification has been performed, the file
ProteinGroups.txt (or ProteinGroups.xlsx) shall be fed to ProtE.

Insert the file path of the ProteinGroups.txt in the `file` parameter.
To copy-paste the file path in Windows, firstly locate the desired file
inside your folders. Hold Shift and right-click the file, then select
“Copy as Path” from the context menu. Go to RStudio and click Ctrl+V or
right-click to paste the path.

Because usually the directories will be separated with a single
backlash, ensure to use forward slashes (/) for specifying paths or
adding a second backlash e.g:

    ProtE_analyse(file = "C:\\Bioprojects\\BreastCancer\\Proteomics\\MaxQuant\\ProteinGroups.txt")

or

    ProtE_analyse(file = "C:/Bioprojects/BreastCancer/Proteomics/MaxQuant/ProteinGroups.txt")

The same logic for the directory separators slashes applies to the
`metadata_file`.

## Setting up `group_names` and number of `samples_per_group`

To enable statistical analysis without the insertion of a metadata file,
in the input file, samples (columns) belonging to the same group must be
sorted next to each other. For example, samples from an experiment with
a 3-groups categorical variable (control, treatment, compound) could be
arranged such that: first columns = Control samples, middle columns =
Treatment samples, last columns = Compound samples.

Group names are defined in the parameter `group_names` as a vector. The
order of the group names inside the vector must follow the order of the
groups by which the samples (columns) have been arranged in the input
proteomics file (from the left to the right). Same goes for the number
of samples of each group, which is defined again as a vector in the
parameter `samples_per_group`. In the following example there are 3
groups (Control,Treatment,Compound) with the Control group consisting of
10 samples the Treatment group of 12 samples and the Compound with 9:

``` r

ProtE_analyse(file = "C:\\Bioprojects\\BreastCancer\\Proteomics\\MaxQuant\\ProteinGroups.txt",group_names = c("Control", "Treatment", "Compound"), samples_per_group = c(10, 12, 9))
```

For the analysis of multiple **Proteome Discoverer** (PD) exports, each
one corresponding to a single sample (case 4) where the input
directories are parsed via the `pd_single_dir` argument, if a
`metadata_file` is not provided, the samples for each experimental Group
of comparison should be inside distinct folders, whose names will be
attributed to each Group.

# Summary of the ProtE pipeline

The pipeline of ProtE is reported in more details in the
`ProtE Workflow` vignette.

## Data Processing

The data are processed via the:

1.  Annotation fetching for the input files that do not include
    established Description

2.  Normalization of proteomic intensity values

3.  Filtering based on the percentage of missing values of each protein.

4.  Imputation of missing data to ensure robust downstream analysis.

## Differential Expression and Enrichment analysis

Once the data processing is complete, the package performs *statistical
analysis* for every pairwise comparison to identify significant protein
abundance differences between experimental groups.

ProtE offers users:

• Conventional statistics (Group averages, Standard Deviations, Log2
Fold Changes )

• Traditional tests for independent and paired samples via the parameter
`independent` Wilcoxon, Kruskal/Friedman, Bartlett & Levene, PERMANOVA

• Limma statistics: Coefficients for every group and covariate,
Anova-like F test results and moderated t-test for every pairwise
comparison. If existing on the `metadata_file`, covariates will also be
fitted in the linear model.

• *GSEA* to find the up and downregulated gene sets/ pathways in every
pairwise comparison.

## Visualization Plots

The results are automatically exported as Excel files, and a range of
visualizations is generated to facilitate quality check and
interpretation. These include:

• Principal Component Analysis (PCA) plots for dimensionality reduction
and group comparison.

• Volcano plots and Heatmaps highlighting significant proteins.

• Protein rank-abundance and meanrank-sd scatterplots.

• Boxplots and violin plots to display data distribution and variability
across groups, both before and after processing

<img src="man/figures/pca_plot.bmp" width="60%" />

<img src="man/figures/enr_plot.bmp" width="60%" />

## Output Directory

The results from each function are saved in a folder named
**ProtE_Analysis**, which is created inside the last directory of the
provided file(s).

ProtE creates 3 sub-folders:

• Data_processing, with the files of the resulting data processing.

• Statistical_Analysis, with the results of the statistical tests and
GSEA.

• Plots, with all the plots saved in .bmp format.

## How to cite ProtE

ProtE was compiled out of R scripts utilized for routine processing,
analysis and visualization of label-free MS data. These scripts were
built and maintained over the course of many years in the Proteomics lab
of Dr Antonia Vlahou, in the BRFAA institute (Greece). To continue
enriching and expanding the functionality of ProtE we kindly request
feedback from the users via contacting <ted.margelos02@gmail.com>

In order to cite ProtE the following reference can be used:
